integration/update_test.go

```diff
@@ -460,3 +461,213 @@ func TestUpdateSet(t *testing.T) {
+					"For example: {$mod: {<field>: ...}} not {$setOnInsert: nan.0}",
+			},
+			alt: "Modifiers operate on fields but we found type double instead. " +
+				"For example: {$mod: {<field>: ...}} not {$setOnInsert: double}",
+		},
+		"err-string": {
+			filter:      bson.D{{"_id", "string"}},
+			setOnInsert: "any string",
+			err: &mongo.WriteError{
+				Code: 9,
+				Message: "Modifiers operate on fields but we found type string instead. " +
+					"For example: {$mod: {<field>: ...}} not {$setOnInsert: \"any string\"}",
+			},
+			alt: "Modifiers operate on fields but we found type string instead. " +
+				"For example: {$mod: {<field>: ...}} not {$setOnInsert: string}",
+		},
+		"err-nil": {
+			filter:      bson.D{{"_id", "nil"}},
+			setOnInsert: nil,
+			err: &mongo.WriteError{
+				Code: 9,
+				Message: "Modifiers operate on fields but we found type null instead. " +
+					"For example: {$mod: {<field>: ...}} not {$setOnInsert: null}",
+			},
+			alt: "Modifiers operate on fields but we found type null instead. " +
+				"For example: {$mod: {<field>: ...}} not {$setOnInsert: null}",
+		},
+	} {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			var err error
+			ctx, collection := setup(t)
+
+			opts := options.Update().SetUpsert(true)
+			var res *mongo.UpdateResult
+			res, err = collection.UpdateOne(ctx, tc.filter, bson.D{{"$setOnInsert", tc.setOnInsert}}, opts)
+			if tc.err != nil {
+				require.Nil(t, tc.res)
+				AssertEqualAltWriteError(t, *tc.err, tc.alt, err)
+				return
+			}
+
+			require.NoError(t, err)
+			id := res.UpsertedID
+			assert.NotEmpty(t, id)
+			res.UpsertedID = nil
+
+			expectedRes := notModified
+			if tc.stat != nil {
+				expectedRes = tc.stat
+			}
+			assert.Equal(t, expectedRes, res)
+
+			var actual bson.D
+			err = collection.FindOne(ctx, tc.filter).Decode(&actual)
+			require.NoError(t, err)
+			AssertEqualDocuments(t, tc.res, actual)
+		})
+	}
+}
+
+func TestSetOnInsertMore(t *testing.T) {
+	t.Parallel()
+
+	notModified := &mongo.UpdateResult{
+		MatchedCount:  0,
+		ModifiedCount: 0,
+		UpsertedCount: 1,
+	}
+
+	for name, tc := range map[string]struct {
+		filter bson.D
+		query  bson.D
+		stat   *mongo.UpdateResult
+		res    bson.D
+		err    *mongo.WriteError
+		alt    string
+	}{
+		"setoninsert-set": {
+			filter: bson.D{{"_id", "test"}},
+			query: bson.D{
+				{"$set", bson.D{{"foo", int32(12)}}},
+				{"$setOnInsert", bson.D{{"value", math.NaN()}}},
+			},
+			res: bson.D{{"_id", "test"}, {"foo", int32(12)}, {"value", math.NaN()}},
+		},
+		"trio": {
+			filter: bson.D{{"_id", "test"}},
+			query: bson.D{
+				{"$set", bson.D{{"foo", int32(12)}}},
+				{"$inc", bson.D{{"foo", int32(1)}}},
+				{"$setOnInsert", bson.D{{"value", math.NaN()}}},
+			},
+			err: &mongo.WriteError{
+				Code:    40,
+				Message: "Updating the path 'foo' would create a conflict at 'foo'",
+			},
+		},
+		"unknown-operator": {
```

Comment:
I suggest extracting this to a separate test as it's not related to `$setOnInsert`. Just like with `TestQueryUnknownFilterOperator`.

Reply:
the entire test set is renamed: it's related to tests when $set, $setOnInsert and $inc are used together.